docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,242 @@
+
+The [Kubernetes cluster autoscaler](https://github.com/kubernetes/autoscaler) currently supports
+scaling on Cluster API deployed clusters for MachineSets and MachineDeployments. One feature
+that is missing from this integration is the ability to scale down to, and up from, a MachineSet
+or MachineDeployment with zero replicas.
+
+This proposal defines an opt-in mechanism whereby Cluster API users and cloud providers can define
+the specific resource requirements for each Infrastructure Machine Template they create. In situations
+where there are zero nodes in the node group, and thus the autoscaler does not have information
+about the nodes, the resource requests are utilized to predict the number of nodes needed. This
+information is only used by the autoscaler when it is scaling from zero.
+
+## Motivation
+
+Allowing the cluster autoscaler to scale down its node groups to zero replicas is a common feature
+implemented for many of the integrated cloud providers. It is a popular feature that has been
+requested for Cluster API on multiple occasions. This feature empowers users to reduce their
+operational resource needs, and likewise reduce their operating costs.
+
+Given that Cluster API is an abstraction point that provides access to multiple concrete cloud
+implementations, this feature might not make sense in all scenarios. To accomodate the wide
+range of deployment options in Cluster API, the scale to zero feature will be optional for
+users and cloud providers.
+
+### Goals
+
+- Provide capability for Cluster API MachineSets and MachineDeployments to scale from and to zero replicas.
+- Create an optional API contract in the Infrastructure Machine Template that allows cloud providers to specify
+  instance resource requirements that will be utilized by the cluster autoscaler.
+- Provide a mechanism for users to override the defined instance resource requirements on any given MachineSet or MachineDeployment.
+
+### Non-Goals/Future Work
+
+- Create an API contract that cloud providers must follow.
+- Create an API that replicates Taint and Label information from Machines to MachineSets and MachineDeployments.
+- Support for MachinePools, either with the cluster autoscaler or using cloud provider native implementations (eg AWS AutoScalingGroups).
+- Create an autoscaling custom resource for Cluster API.
+- Add MachinePool support to the cluster autoscaler.
+
+## Proposal
+
+To facilitate scaling from zero replicas, the minimal information needed by the cluster autoscaler
+is the CPU and memory resources for the target node group that will be scaled. The autoscaler uses
+this information to create a prediction about how many nodes should be created when scaling. In
+most situations this information can be directly read from the nodes that are running within a
+node group. But, during a scale from zero situation (ie when a node group has zero replicas) the
+autoscaler needs to acquire this information from the cloud provider.
+
```

Comment:
@fabriziopandini @elmiko 

Maybe an idea to address the concern mentioned by Fabrizio here: https://github.com/kubernetes-sigs/cluster-api/pull/4283#r593444976

What about only using the annotations on both the InfrastructureMachineTemplates and the MachineDeployment, ....

Possible advantages:
* CAPI types do not have to be changed for this, so it would be basically a feature build on top of CAPI APIs by autoscaler via annotations
* Only autoscaler implementation has to be adjusted
* Because of this it would work with all infrastructure providers out of the box (automatically reconciling the data on InfrastructureMachineTemplates could still be implemented, but it would also work without any changes to any InfraStructureProvider)

I think that would be a good option to address the concern that we spread autoscaler concerns into CAPI APIs.

Reply:
this is certainly an approach we could take, i had avoided going down this path further as i had been advised that annotations are a very weak form of API.

this is how we do scale from zero in openshift today, we use an annotation on the machineset to inform the autoscaler about the sizing requirements. i would be happy to contribute this solution, i think it could be even easier than suggested above. we could allow annotations to exist on either a machinedeployment or machineset, no need for the infrastructure template (as it would require another call to the api server from inside the autoscaler).

Reply:
I think if we have those fields/annotations only MachineDeployments/MachineSets or also on InfrastructureMachineTemplates is something which can be decided independently from if we use annotations or fields.

So what I was proposing was to have annotations on MachineDeployments/MachineSets and InfrastructureMachineTemplates (if still desired).

I agree with that the annotations are a weaker form of API (mostly because they are not as type safe as regular fields), but to me it feels like it would be an ideal way to layer the autoscaler API on top of the CAPI types. Let's see what others think about it.

Reply:
yeah, i'm totally fine with this approach and it makes it very easy since we already have this implemented for openshift. slightly tangential, but i would like to see us create a new resource that can be used for cluster autoscaling, so maybe in the end it will be better for us to use a lightweight solution now. 

thanks for the feedback @sbueringer , i have added this as a topic for the meeting today. maybe we can come to a consensus.